blues/reporters.py

```diff
@@ -0,0 +1,538 @@
+        self._parameters = parameters
+
+    def report(self, simulation, state):
+        """Generate a report.
+        Parameters
+        ----------
+        simulation : simtk.openmm.app.Simulation
+            The Simulation to generate a report for
+        state : simtk.openmm.State
+            The current state of the simulation
+        """
+        if not self._is_intialized:
+            self._initialize(simulation)
+            self._is_intialized = True
+
+        self._checkForErrors(simulation, state)
+
+        args = ()
+        kwargs = {}
+        if self._coordinates:
+            coordinates = state.getPositions(asNumpy=True)[self._atomSlice]
+            coordinates = coordinates.value_in_unit(getattr(units, self._traj_file.distance_unit))
+            if self._frame_indices:
+                if simulation.currentStep not in self._frame_indices:
+                    coordinates = np.zeros(coordinates.shape)
+            args = (coordinates,)
+        if self._time:
+            kwargs['time'] = state.getTime()
+        if self._cell:
+            vectors = state.getPeriodicBoxVectors(asNumpy=True)
+            vectors = vectors.value_in_unit(getattr(units, self._traj_file.distance_unit))
+            a, b, c, alpha, beta, gamma = unitcell.box_vectors_to_lengths_and_angles(*vectors)
+            kwargs['cell_lengths'] = np.array([a, b, c])
+            kwargs['cell_angles'] = np.array([alpha, beta, gamma])
+        if self._potentialEnergy:
+            kwargs['potentialEnergy'] = state.getPotentialEnergy()
+        if self._kineticEnergy:
+            kwargs['kineticEnergy'] = state.getKineticEnergy()
+        if self._temperature:
+            kwargs['temperature'] = 2*state.getKineticEnergy()/(self._dof*units.MOLAR_GAS_CONSTANT_R)
+        if self._velocities:
+            kwargs['velocities'] = state.getVelocities(asNumpy=True)[self._atomSlice, :]
+
+        #add a portion like this to store things other than the protocol work
+        if self._protocolWork:
+            protocol_work = simulation.integrator.get_protocol_work(dimensionless=True)
+            kwargs['protocolWork'] = np.array([protocol_work])
+        if self._alchemicalLambda:
+            kwargs['alchemicalLambda'] = np.array([simulation.integrator.getGlobalVariableByName('lambda')])
+        if self._title:
+            kwargs['title'] = self._title
+        if self._parameters:
+            kwargs['parameters'] = self._parameters
+        if self._environment:
+            kwargs['environment'] = self._environment
+
+        self._traj_file.write(*args, **kwargs)
+        # flush the file to disk. it might not be necessary to do this every
+        # report, but this is the most proactive solution. We don't want to
+        # accumulate a lot of data in memory only to find out, at the very
+        # end of the run, that there wasn't enough space on disk to hold the
+        # data.
+        if hasattr(self._traj_file, 'flush'):
```

Comment:
We might not want to flush the data to the disk on each report (maybe only at the end of an iteration), since writing to a file is relatively slow. It might be worthwhile to do some benchmarking to see how often flushing affects simulation speeds.

Reply:
I think I'll keep this around since this is what is found in the native reporter from MDTraj.

Reply:
You are right! Flushing to disk doubles the timing.

Simulation parameters:
```
 opt = { 'temperature' : 300.0, 'friction' : 1, 'dt' : 0.002,
            'nIter' : 50, 'nstepsNC' : 100, 'nstepsMD' : 100, 'nprop' : 1,
            'nonbondedMethod' : 'PME', 'nonbondedCutoff': 10,
            'constraints': 'HBonds', 'freeze_distance' : 0.0,
            'trajectory_interval' : 10, 'reporter_interval' : 10,
            'write_move' : False,
            'platform' : platform_name,
            'outfname' : 'tolwater'}
```

**HDF5 Reporter with flushing at each step**
> real	0m41.297s
user	0m38.294s
sys	0m3.396s

**HDF5 Reporter without flushing at each step**
> real	0m18.254s
user	0m16.388s
sys	0m1.896s

**Without HDF5 Reporter**
>real	0m18.142s
user	0m16.339s
sys	0m1.842s